### PR TITLE
FEATURE: introduces Discourse Lightbox

### PR DESCRIPTION
**Pull request by hnb-ku**
Please see https://meta.discourse.org/t/migrating-away-from-magnific-popup/251505 for more details about this PR

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/initializers/chat-decorators.js [hunk 1]
```diff
@@ -63,13 +65,40 @@ export default {
...
+        // listener to the parent element (chat message in this case)
+        //
+        // Topics have a cleanupStream method that is called when the topic is removed
+        // api.cleanupStream(cleanUpLightboxes);
+        //
+        // The implmentation would look something like this
+        // api.cleanupChatStream(lightboxService.cleanupLightboxes);
+        //
+        // that said, the new lightbox event listners are cleaned up in dom:clean which
+        // fires on route transition anyway.
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/initializers/chat-decorators.js:91]: Chat doesn't have cloaking for now, all messages stay in DOM so there's no real need for such API at the moment.

I don't think it would leaking when changing channel, but worth checking.

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+      passive: true,
+      capture: true,
+    });
+
+    document.activeElement.blur();
+
+    this.lastFocusedElement?.focus();
+  }
+
+  async #reset() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:253]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+      }
+    };
+
+    document.addEventListener("focus", this.onFocus, {
+      passive: true,
+      capture: true,
+    });
+  }
+
+  async #restoreDocumentFocus() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:242]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    const lastFocusableElement =
+      focusableElements[focusableElements.length - 1];
+
+    const focusTraps = document.querySelectorAll("[data-lightbox-focus-trap]");
+
+    const firstfocusTrap = focusTraps[0];
+    const lastfocusTrap = focusTraps[focusTraps.length - 1];
+
+    this.onFocus = ({ target }) => {
+      if (target === firstfocusTrap) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:229]: At this point in time you are not sure the elements still exist here, most likely yes but still it will be happening in a different context

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+      handlerOptions,
+    });
+  }
+
+  @bind
+  async setupLightboxes({ container, selector }) {
+    this.#setupLightboxes({ container, selector });
+  }
+
+  async #cleanupLightboxes() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:193]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+      ({ container, clickHandler, handlerOptions }) => {
+        container.removeEventListener("click", clickHandler, handlerOptions);
+      }
+    );
+
+    this.lightboxClickElements = [];
+  }
+
+  @bind
+  async cleanupLightboxes() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:206]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+    if (!hasLightboxes) {
+      return;
+    }
+
+    const handlerOptions = {
+      capture: true,
+    };
+
+    function clickHandler(event) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:151]: Should probably be a dedicated function

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+      startingIndex,
+      callbacks: { ...this.callbacks },
+      options: { ...this.options },
+    });
+
+    this.lightboxIsOpen = true;
+  }
+
+  @bind
+  async closeLightbox() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:124]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    if (this.lightboxIsOpen) {
+      this.appEvents.trigger(LIGHTBOX_APP_EVENT_NAMES.CLOSE);
+      this.lightboxIsOpen = false;
+    }
+  }
+
+  async #setupLightboxes({ container, selector }) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:131]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+  }
+
+  @bind
+  onLightboxCleanedUp() {
+    return true;
+  }
+
+  @bind
+  async openLightbox({ container, selector }) {
+    const { items, startingIndex } = await processHTML({ container, selector });
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:111]: What if there's an error in processHTML?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+  @bind
+  async onLightboxItemDidChange({ currentItem }) {
+    this.appEvents.trigger(LIGHTBOX_APP_EVENT_NAMES.ITEM_DID_CHANGE, {
+      currentItem,
+    });
+  }
+
+  @bind
+  async onLightboxClosed() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:90]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+  @bind
+  async onLightboxItemWillChange({ currentItem }) {
+    this.appEvents.trigger(LIGHTBOX_APP_EVENT_NAMES.ITEM_WILL_CHANGE, {
+      currentItem,
+    });
+  }
+
+  @bind
+  async onLightboxItemDidChange({ currentItem }) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:83]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+    this.appEvents.on(
+      LIGHTBOX_APP_EVENT_NAMES.CLEAN,
+      this,
+      this.cleanupLightboxes
+    );
+  }
+
+  @bind
+  async onLightboxOpened({ items, currentItem }) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:60]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+    this.#setupDocumentFocus();
+
+    this.appEvents.trigger(LIGHTBOX_APP_EVENT_NAMES.OPENED, {
+      items,
+      currentItem,
+    });
+  }
+
+  @bind
+  async onLightboxItemWillChange({ currentItem }) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:76]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+  lightboxClickElements = [];
+  lastFocusedElement = null;
+  originalSiteThemeColor = null;
+  onFocus = null;
+  onLightboxOpenerClick = null;
+
+  callbacks = {};
+  options = {};
+
+  async init() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:32]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+    this.originalSiteThemeColor = "";
+    this.lightboxIsOpen = false;
+
+    this.appEvents.trigger(LIGHTBOX_APP_EVENT_NAMES.CLOSED);
+  }
+
+  @bind
+  onLightboxCleanedUp() {
+    return true;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:106]: do we check the return value of this somewhere?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+    this.lightboxClickElements = [];
+  }
+
+  @bind
+  async cleanupLightboxes() {
+    this.#cleanupLightboxes();
+  }
+
+  async #setupDocumentFocus() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:210]: What is the reason for this to be async?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+
+  async #setupLightboxes({ container, selector }) {
+    if (!container) {
+      throw new Error("Lightboxes require a container to be passed in");
+    }
+
+    if (container.length) {
+      // passed in a nodelist instead of an element
+      container = findNearestSharedParent(container);
+    }
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:139]: Why do we allow this?

File: app/assets/javascripts/discourse/app/lib/lightbox/helpers/site-theme-color.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+import { postRNWebviewMessage } from "discourse/lib/utilities";
+
+export async function getSiteThemeColor() {
+  const siteThemeColor = document.querySelector('meta[name="theme-color"]');
+  return siteThemeColor?.content || null;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/lightbox/helpers/site-theme-color.js:5]:
```suggestion
  return siteThemeColor?.content;
```

File: app/assets/javascripts/discourse/app/components/uppy-image-uploader.js [hunk 20]
```diff
@@ -97,13 +106,24 @@ export default Component.extend(UppyUploadMixin, {
 
   @on("didRender")
   _applyLightbox() {
-    next(() => lightbox(this.element, this.siteSettings));
+    if (this.experimentalLightboxEnabled) {
+      setupLightboxes({
+        container: this.element,
+        selector: ".lightbox",
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/uppy-image-uploader.js:112]: Maybe this should just be the default?

File: app/assets/javascripts/discourse/app/components/d-lightbox/screen-reader-announcer.hbs [hunk 21]
```diff
@@ -0,0 +1,21 @@
+<div
+  data-lightbox-screen-reader-announcer
+  class="d-lightbox__content__screen-reader-announcer"
+  tabindex="-1"
+>
+  <h2
+    data-lightbox-screen-reader-announcer-title
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/screen-reader-announcer.hbs:7]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/screen-reader-announcer.hbs [hunk 22]
```diff
@@ -0,0 +1,21 @@
+<div
+  data-lightbox-screen-reader-announcer
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/screen-reader-announcer.hbs:2]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+        data-lightbox-fullscreen-button
+        aria-hidden="true"
+        @class="d-lightbox__content__header__buttons__full-screen-button"
+        @name="fullscreen"
+        @icon="discourse-expand"
+        @onClick={{@toggleFullScreen}}
+      />
+    {{/if}}
+    <DLightbox::Button
+      data-lightbox-close-button
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:66]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+        aria-hidden="true"
+        @class="d-lightbox__content__header__buttons__new-tab-button"
+        @name="newtab"
+        @icon="external-link-alt"
+        @onClick={{@openInNewTab}}
+      />
+    {{/if}}
+    {{#if @canFullscreen}}
+      <DLightbox::Button
+        data-lightbox-fullscreen-button
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:57]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+          data-lightbox-counter-total
+          class="d-lightbox__content__header__multi-item-controls__counter__total"
+        >
+          {{~@totalItemCount~}}
+        </span>
+      </div>
+    </div>
+  {{/if}}
+  <div
+    data-lightbox-header-buttons
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:42]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+          {{~@counterIndex~}}
+        </span>
+        <span
+          data-lightbox-counter-separator
+          class="d-lightbox__content__header__multi-item-controls__counter__separator"
+        >
+          <span>/</span>
+        </span>
+        <span
+          data-lightbox-counter-total
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:33]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+        class="d-lightbox__content__header__multi-item-controls__counter"
+      >
+        <span
+          data-lightbox-counter-current
+          class="d-lightbox__content__header__multi-item-controls__counter__current"
+        >
+          {{~@counterIndex~}}
+        </span>
+        <span
+          data-lightbox-counter-separator
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:27]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+        @name="carousel"
+        @icon="images"
+        @onClick={{@toggleCarousel}}
+      />
+      <div
+        data-lightbox-counter
+        class="d-lightbox__content__header__multi-item-controls__counter"
+      >
+        <span
+          data-lightbox-counter-current
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:21]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 29]
```diff
@@ -0,0 +1,73 @@
+<div data-lightbox-header class="d-lightbox__content__header">
+  {{#if @canNavigate}}
+    <div
+      data-lightbox-multi-item-controls
+      aria-hidden="true"
+      class="d-lightbox__content__header__multi-item-controls"
+    >
+      <DLightbox::Button
+        data-lightbox-carousel-button
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:9]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 30]
```diff
@@ -0,0 +1,73 @@
+<div data-lightbox-header class="d-lightbox__content__header">
+  {{#if @canNavigate}}
+    <div
+      data-lightbox-multi-item-controls
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:4]: why do we need data attribute here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs [hunk 31]
```diff
@@ -0,0 +1,73 @@
+<div data-lightbox-header class="d-lightbox__content__header">
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/header.hbs:1]: why do we need data attribute here?

**Review by jjaffeux — COMMENTED** [review]
Great to have this gone :+1: I didn't put it in comments as it was everywhere but I have two points:
- Marking async something that does only synchronous code shouldn't be done
- We shouldn't use data-attributes with no values for targeting. Adding/Removing could be fine. But just basically having data-attributes which are the duplicates of css class seems very much overkill and will create a lot of work for any change. It's also not really a concern here, but targeting by data attribute is not performant

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 32]
```diff
@@ -0,0 +1,465 @@
...
+        break;
+      case SWIPE_DIRECTIONS.DOWN:
+        this.toggleCarousel();
+        break;
+    }
+  }
+
+  @bind
+  cleanup() {
+    this.hasCarousel = false;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:438]: Component could be destroyed at this point in time

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 33]
```diff
@@ -0,0 +1,465 @@
...
+    discourseLater(this.cleanup, this.animationDuration);
+
+    this.callbacks.onClose?.();
+  }
+
+  async #setCurrentItem(index) {
+    await this.#onBeforeItemChange();
+
+    this.currentIndex = (index + this.totalItemCount) % this.totalItemCount;
+    this.currentItem = await preloadItemImages(this.items[this.currentIndex]);
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:228]: what if we have an error here?

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 34]
```diff
@@ -0,0 +1,465 @@
...
+  async close() {
+    this.willClose = true;
+
+    discourseLater(this.cleanup, this.animationDuration);
+
+    this.callbacks.onClose?.();
+  }
+
+  async #setCurrentItem(index) {
+    await this.#onBeforeItemChange();
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:225]: what if we have an error here?

File: app/assets/javascripts/discourse/app/components/d-lightbox/footer.hbs [hunk 35]
```diff
@@ -0,0 +1,55 @@
...
+        @class="d-lightbox__content__footer__buttons__rotate-button"
+        @name="rotate"
+        @icon="redo"
+        @onClick={{@rotateImage}}
+      />
+    {{/if}}
+    {{#if @canDownload}}
+      <DLightbox::Button
+        data-lightbox-download-button
+        @class="d-lightbox__content__footer__buttons__download-button"
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/footer.hbs:48]: @chapoi could you advise on BEM on the PR please? I don't think we should have such long names

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 36]
```diff
@@ -0,0 +1,465 @@
...
+      `${base}-rotation: ${this.rotationAmount}deg`,
+      `${base}-width: ${width}px`,
+      `${base}-height: ${height}px`,
+      `${base}-aspect-ratio: ${aspectRatio}`,
+      `${base}-dominant-color: #${dominantColor}`,
+      `${base}-full-size-url: url(${fullsizeURL})`,
+      `${base}-small-url: url(${smallURL})`
+    );
+
+    return htmlSafe(variables.filter(Boolean).join(";"));
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:83]: Do we need to escape the URLs here to make sure it can't be used an an XSS vector?

**Comment by jjaffeux** [comment]
The escape shortcut should work all the time IMO, ATM it works only if you have the [x] in focus.

**Comment by jjaffeux** [comment]
I wonder if we couldn't start zoomed right away if we know the image is bigger than viewport?

File: app/assets/javascripts/discourse/app/components/d-lightbox/footer.hbs (hunk 35 above)
Comment (chapoi) [reply app/assets/javascripts/discourse/app/components/d-lightbox/footer.hbs:48]: Yup, since there is only 1 download-button in the whole component (I think), this should be enough to identify it

```suggestion
        @class="d-lightbox__download-button"
```

**Comment by SamSaffron** [comment]
@hnb-ku let us know when you are ready for another round of reviews, so excited about this change.

File: plugins/chat/assets/javascripts/discourse/initializers/chat-decorators.js (hunk 1 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/initializers/chat-decorators.js:93]: yes sure remove

File: .gitignore [hunk 37]
```diff
@@ -52,6 +52,7 @@
 
 # We provide a .sample but people can use newer versions if they want to
 .ruby-version
+.ruby-gemset
```
Comment (jjaffeux) [inline .gitignore:55]: If you want to add this, should probably be in another PR

**Comment by jjaffeux** [comment]
So a general design feedback is that maybe `<DLightbox />` main component in `application.hbs` should be piloted through a service instead of AppEvents.


```javascript
const lightbox = getOwner(target).lookup("service:lightbox")
lightbox.items = [item1, item2, ...];
lightbox.open({currentItem: item2});
lightbox.close()
```

And then `<DLightbox />` component would only react to this state

```
class DLightbox extends Component {
  @service lightbox;
  
  get shouldRender() {
    return this.lightbox.opened && this.lightbox.items.length; // or encapsulate this logic directly in lightbox service
  }
  
  get items() {
    return this.lightbox.items;
  }
}

<template>
  {{#each this.items as |item|}}
    ...
  {{/each}}
</tempalte>
```

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 38]
```diff
@@ -0,0 +1,476 @@
...
+      this.isRotated && `${base}--is-rotated`,
+      this.canZoom && `${base}--can-zoom`,
+      this.hasExpandedTitle && `${base}--has-expanded-title`,
+      this.hasCarousel && `${base}--has-carousel`,
+      this.hasLoadingError && `${base}--has-loading-error`,
+      this.willClose && `${base}--will-close`,
+      this.isRotated &&
+        this.rotationAmount &&
+        `${base}--is-rotated-${this.rotationAmount}`
+    );
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:110]: Can you look CSS naming is correct for you please @chapoi ?

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 39]
```diff
@@ -0,0 +1,266 @@
...
+    this.callbacks = {
+      onOpen: this.onLightboxOpened,
+      onClose: this.onLightboxClosed,
+      onWillChange: this.onLightboxItemWillChange,
+      onItemDidChange: this.onLightboxItemDidChange,
+      onCleanUp: this.onLightboxCleanedUp,
+    };
+
+    this.options = {
+      isMobile: Mobile.mobileView,
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:43]: why does it need to be an option? We should also just use the service. (@service site; this.site.mobileView)

File: app/assets/javascripts/discourse/app/components/d-lightbox/body.hbs [hunk 40]
```diff
@@ -0,0 +1,57 @@
+<div
+  class="d-lightbox__body"
+  tabindex="-1"
+  {{on "touchstart" @onTouchstart passive=true capture=true}}
+  {{on "touchend" @onTouchend passive=true capture=true}}
+  {{on "click" @toggleZoom passive=true}}
+>
+  {{#if @shouldDisplayMainImageArrows}}
+    <DLightbox::Button
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/body.hbs:9]: Could you justify why we need these custom buttons please? It would be better to rely on our own DButton and improve them if necessary

File: app/assets/javascripts/discourse/app/components/d-lightbox/body.hbs [hunk 41]
```diff
@@ -0,0 +1,57 @@
+<div
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/body.hbs:1]: Another question I have, could have this been made in a `<DModal />` ? Instead of having to rebuild all the backdrop logic and other things.

File: app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs [hunk 42]
```diff
@@ -0,0 +1,47 @@
...
+        fetchPriority="low"
+        decoding="async"
+        loading="lazy"
+        tabindex="-1"
+        src={{item.smallURL}}
+        data-lightbox-item-index={{item.index}}
+        style={{item.cssVars}}
+        class={{concat
+          "d-lightbox__carousel-item"
+          (if (eq item.index @currentIndex) "--is-current active")
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs:34]: should probably be `-active` and `is-current` here @chapoi ?

File: app/assets/javascripts/discourse/app/components/d-lightbox/screen-reader-announcer.hbs [hunk 43]
```diff
@@ -0,0 +1,16 @@
+<div class="d-lightbox__screen-reader-announcer" tabindex="-1">
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/screen-reader-announcer.hbs:1]: That's really nice

File: app/assets/javascripts/discourse/app/components/uppy-image-uploader.hbs [hunk 44]
```diff
@@ -25,15 +25,25 @@
...
+          @disabled={{this.loadingLightbox}}
+          data-lightbox-trigger
+        />
+      {{else}}
+        <DButton
+          @icon="discourse-expand"
+          @title="expand"
+          @type="button"
+          @class="image-uploader-lightbox-btn no-text"
+          @action={{action "toggleLightbox"}}
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/uppy-image-uploader.hbs:43]:
```suggestion
          @action={{this.toggleLightbox}}
```

File: app/assets/javascripts/discourse/app/lib/lightbox/constants.js [hunk 45]
```diff
@@ -0,0 +1,81 @@
+import { isTesting } from "discourse-common/config/environment";
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/lightbox/constants.js:1]: that's an interesting pattern :+1:

File: app/assets/javascripts/discourse/app/services/lightbox.js [hunk 46]
```diff
@@ -0,0 +1,266 @@
...
+    );
+
+    this.lightboxClickElements = null;
+    this.lastFocusedElement = null;
+    this.originalSiteThemeColor = null;
+    this.lightboxIsOpen = null;
+    this.onFocus = null;
+    this.callbacks = null;
+    this.options = null;
+    this.selector = null;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/services/lightbox.js:260]: Is that needed?

File: app/assets/javascripts/discourse/app/components/d-lightbox.js (hunk 38 above)
Comment (chapoi) [reply app/assets/javascripts/discourse/app/components/d-lightbox.js:110]: Okay so, the idea is that modifiers are standalone classes (not chained), with either `-` as prefix or `has-/is-`, in case of temporary/changable states.

I think based on naming this should be mostly states? So:

```suggestion
    classNames.push(
      this.layoutType && `is-${this.layoutType}`,
      this.isVisible && `is-visible`,
      this.isLoading ? `is-loading` : `is-finished-loading`,
      this.isFullScreen && `is-fullscreen`,
      this.isZoomed && `is-zoomed`,
      this.isRotated && `is-rotated`,
      this.canZoom && `can-zoom`,
      this.hasExpandedTitle && `has-expanded-title`,
      this.hasCarousel && `has-carousel`,
      this.hasLoadingError && `has-loading-error`,
      this.willClose && `will-close`,
      this.isRotated &&
        this.rotationAmount &&
        `is-rotated-${this.rotationAmount}`
    );
```

File: app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs [hunk 47]
```diff
@@ -1,10 +1,10 @@
 <div class="d-lightbox__carousel">
   {{#if @shouldDisplayCarouselArrows}}
-    <DLightbox::Button
-      @class="d-lightbox__carousel-previous-button"
-      @name="previous"
+    <D::Button
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs:3]: Is github tricking me? it seems like it shows "D::Button" and should be "DButton"

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 48]
```diff
@@ -0,0 +1,476 @@
...
+import { htmlSafe } from "@ember/template";
+import { inject as service } from "@ember/service";
+import { tracked } from "@glimmer/tracking";
+
+export default class DLightbox extends Component {
+  @service appEvents;
+
+  @tracked callbacks = {};
+  @tracked items = [];
+  @tracked options = {};
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:32]: I don't need it needs to be tracked, it's only getting initial state when opening the lightbox, but not changed afterwards.

Might be the same for items and callbacks.

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 49]
```diff
@@ -0,0 +1,476 @@
...
+
+  @tracked isVisible = false;
+  @tracked isLoading = false;
+  @tracked willClose = false;
+
+  @tracked currentIndex = 0;
+  @tracked currentItem = {};
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:39]: should just be `@tracked currentItem;`

You have it or you don't have it.

File: app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs (hunk 47 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs:3]: No worries, and yes I guess both can work

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 50]
```diff
@@ -0,0 +1,476 @@
...
+  get previousButtonIcon() {
+    return this.options.isRTL ? "chevron-right" : "chevron-left";
+  }
+
+  get zoomButtonIcon() {
+    return this.isZoomed ? "search-minus" : "search-plus";
+  }
+
+  @bind
+  async registerAppEventListeners() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:185]: don't think it needs to be async there's no blocking calls inside

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 51]
```diff
@@ -0,0 +1,476 @@
...
+    this.appEvents.on(LIGHTBOX_APP_EVENT_NAMES.OPEN, this.open);
+    this.appEvents.on(LIGHTBOX_APP_EVENT_NAMES.CLOSE, this.close);
+  }
+
+  @bind
+  async deregisterAppEventListners() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:191]: don't think it needs to be async there's no blocking calls inside

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 52]
```diff
@@ -0,0 +1,476 @@
...
+    }
+
+    this.callbacks.onOpen?.({
+      items: this.items,
+      currentItem: this.currentItem,
+    });
+  }
+
+  @bind
+  async close() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:224]: don't think it needs to be async there's no blocking calls inside

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 53]
```diff
@@ -0,0 +1,476 @@
...
+    this.willClose = true;
+
+    discourseLater(this.cleanup, this.animationDuration);
+
+    this.callbacks.onClose?.();
+  }
+
+  async #setCurrentItem(index) {
+    await this.#onBeforeItemChange();
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:233]: we are not waiting on anything inside #onBeforeItemChange so it doesn’t need to await here, and also doesn’t need to be `async #onBeforeItemChange`

File: app/assets/javascripts/discourse/app/components/d-lightbox.js [hunk 54]
```diff
@@ -0,0 +1,476 @@
...
+      behavior: "smooth",
+    });
+
+    this.callbacks.onItemDidChange?.({
+      currentItem: this.currentItem,
+    });
+
+    const nextItem = this.items[this.currentIndex + 1];
+    return nextItem ? preloadItemImages(nextItem) : false;
+  }
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-lightbox.js:266]: same nothing is waiting here, so the whole function being async is useless

File: app/assets/javascripts/discourse/app/components/d-lightbox.js (hunk 49 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/components/d-lightbox.js:39]: Yes currentIndex is most likely needed indeed, I was talking of `currentItem` here

File: app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs (hunk 42 above)
Comment (chapoi) [reply app/assets/javascripts/discourse/app/components/d-lightbox/carousel.hbs:35]: Not sure how `current` is used. If it's simply a different static styling do `-current`. Whats the difference between current and active?

File: app/assets/javascripts/discourse/app/components/d-lightbox.js (hunk 53 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/components/d-lightbox.js]: Yes you right, I missed that one, that seems correct

**Comment by chapoi** [comment]
One general design feedback: we don't use this kind of tertiary highlight on a button anywhere in Discourse as far as I know:
image

It's either boxed:
image

Or highlight with greyscale:
image

image


So I wonder if we should try to be more similar here? 
@jordanvidrine @awesomerobot What do you think? Too nitpicky?

**Review by jjaffeux — APPROVED** [review]
We need to get this out, thanks for the heroic work to move it to the finish line @dbattersby 

I do think there is much more work to do in the future as hinted in the review, but we have already been sitting on this for far too long, also it's behind an experimental flag so we can iterate.

**Comment by nattsw** [comment]
Congrats!!
